### PR TITLE
port f4 gpios

### PR DESCRIPTION
**Pull request by burrbull**

### CODE REVIEW

**Review by richardeoin — APPROVED** [review]
This is an impressive piece of work, thank you!

I tried out this branch with a reasonably complex project. It needed relatively little refactoring, and I don't see any regressions. It solves several pain-points in GPIO usage and keeps this crate in line with the F4 HAL also.

File: src/gpio/convert.rs [hunk 1]
```diff
@@ -0,0 +1,731 @@
...
+    }
+}
+
+macro_rules! af {
+    ($($into_alternate_af:ident: $A:literal;)+) => {
+        $(
+            #[doc="Configures the pin to operate in alterante function "]
+            #[doc=stringify!($A)]
+            #[doc=" mode"]
+            #[deprecated(since = "0.12.0")]
```
[the remark this answers is not in the document]
Comment (mattico) [reply src/gpio/convert.rs:378]: > `#[doc="Configures the pin to operate in alterante function "]`

Typo: alterante

**Review by mattico — APPROVED** [review]
I also tested this out and like it. Should we replace all uses of `Never` with `Infallible`?

**Comment by richardeoin** [comment]
@mattico Yes, but let's do that in another PR

**Comment by richardeoin** [comment]
bors cancel

**Comment by richardeoin** [comment]
This is failing because the check on latest stable [was removed from the CI run](https://github.com/stm32-rs/stm32h7xx-hal/pull/334/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fL15). I assume this is a mistake, let's try to fix it

**Comment by richardeoin** [comment]
bors retry

**Comment by richardeoin** [comment]
bors retry

**Pull request merged**